File: C++/Solution502.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <unordered_map>
#include <unordered_set>
#include <queue>
#include <stack>
using namespace std;

typedef long long ll;
typedef pair<int, int> pii;

class Solution502 {
public:
    int findMaximizedCapital(int k, int w, vector<int>& profits, vector<int>& capital) {
        vector<pii> items(profits.size());
        for(int i = 0; i < profits.size(); i++){
            items[i] = {capital[i], profits[i]};
        }
        int size = profits.size();
        sort(items.begin(), items.end(), less<pii>());
        priority_queue<int, vector<int>, less<int>> pq;
        int index = 0;
        while(k > 0){
            while(index < size && items[index].first <= w){
                pq.push(items[index].second);
                index ++;
            }
            if(pq.empty()){
                return w;
            }
            w += pq.top();
            pq.pop();
            k --;
        }
        return w;
    }
};

File: C++/Solution502.java
import java.util.*;

class Solution502 {
    public int findMaximizedCapital(int k, int w, int[] profits, int[] capital) {
        int size = profits.length;
        int[][] items = new int[size][2];
        for(int i = 0; i < size; i++){
            items[i][0] = capital[i];
            items[i][1] = profits[i];
        }
        Arrays.sort(items, (a, b) -> a[0] - b[0]);
        int index = 0;
        Queue<Integer> pq = new PriorityQueue<>(
            new Comparator<>(){
                @Override
                public int compare(Integer x, Integer y){
                    return y - x;
                }
            }
        );
        while(k > 0){
            while(index < size && items[index][0] <= w){
                pq.offer(items[index][1]);
                index ++;
            }
            if(pq.isEmpty()){
                return w;
            }
            w += pq.poll();
            k --;
        }
        return w;
    }

    public static void main(String[] args){
        Solution502 s = new Solution502();
        s.findMaximizedCapital(2, 2, new int[]{0, 1}, new int[]{1, 0});
    }
}